File: Conversoes de Bases/Decimal e Hexadecimal/DecimalHexadecimal.py
import math

LimiteMaximoDecimal = 2147483647
LimiteMaximoHexadecimal = '7FFFFFFF'

def DecimalParaHexadecimal(decimal):
    hexadecimal = ""
    while decimal > 0:
        resto = decimal % 16
        if resto <= 9:
            hexadecimal = str(resto) + hexadecimal
        else:
            hexadecimal = chr(resto - 10 + ord('A')) + hexadecimal
        decimal = decimal // 16
    return hexadecimal

def HexadecimalParaDecimal(hexadecimal):
    decimal = 0
    potencia = 0
    while len(hexadecimal) > 0:
        digito = hexadecimal[-1].upper()
        if digito.isdigit():
            decimal += int(digito) * math.pow(16, potencia)
        else:
            decimal += (ord(digito) - ord('A') + 10) * math.pow(16, potencia)
        hexadecimal = hexadecimal[:-1]
        potencia += 1
    return int(decimal)

def IsValidHexadecimal(hexadecimal):
    for digito in hexadecimal:
        if not (digito.isdigit() or digito.upper() in ['A', 'B', 'C', 'D', 'E', 'F']):
            return False
    return True

opcao = -1
while opcao != 0:
    print('===============================')
    print(' CONVERSOR DECIMAL-HEXADECIMAL ')
    print('===============================')
    print()
    print('1 - Converter decimal para hexadecimal')
    print('2 - Converter hexadecimal para decimal')
    print('0 - Sair')
    print()
    opcao = int(input('Digite a opcao desejada: '))
    if opcao == 1:
        print()
        decimal = int(input(f'Digite um numero decimal (ate {LimiteMaximoDecimal}): '))
        if decimal > LimiteMaximoDecimal:
            print('Valor decimal fora do limite permitido.')
        else:
            print(f'Valor em hexadecimal: {DecimalParaHexadecimal(decimal)}')
    elif opcao == 2:
        print()
        hexadecimal = input(f'Digite um numero hexadecimal (ate {LimiteMaximoHexadecimal}): ')
        if not IsValidHexadecimal(hexadecimal):
            print('Valor hexadecimal invalido.')
        elif HexadecimalParaDecimal(hexadecimal) > LimiteMaximoDecimal:
            print('Valor hexadecimal fora do limite permitido.')
        else:
            print(f'Valor em decimal: {HexadecimalParaDecimal(hexadecimal)}')
    print()

File: Conversoes de Bases/Decimal e Hexadecimal/DecimalHexadecimal.scala
import scala.io.StdIn.readLine
import scala.math.pow

object Main { // DecimalHezadecimal
  val LimiteMaximoDecimal: Long = 2147483647
  val LimiteMaximoHexadecimal: String = "7FFFFFFF"
  def DecimalParaHexadecimal(decimal: Long): String = {
    var resto: Long = 0
    var hexadecimal: String = ""
    var tempDecimal = decimal
    while (tempDecimal > 0) {
      resto = tempDecimal % 16
      resto match {
        case r if r >= 0 && r <= 9 =>
          hexadecimal = (r + '0').toChar.toString + hexadecimal
        case r if r >= 10 && r <= 15 =>
          hexadecimal = (r - 10 + 'A').toChar.toString + hexadecimal
      }
      tempDecimal = tempDecimal / 16
    }
    hexadecimal
  }

  def HexadecimalParaDecimal(hexadecimal: String): Long = {
    var potencia: Long = 0
    var digito: Char = ' '
    var decimal: Long = 0
    var tempHexadecimal = hexadecimal
    while (tempHexadecimal.nonEmpty) {
      digito = tempHexadecimal.last.toUpper
      digito match {
        case d if d >= '0' && d <= '9' =>
          decimal += (d - '0') * pow(16, potencia).toLong
        case d if d >= 'A' && d <= 'F' =>
          decimal += (d - 'A' + 10) * pow(16, potencia).toLong
      }
      tempHexadecimal = tempHexadecimal.dropRight(1)
      potencia += 1
    }
    decimal
  }

  def IsValidHexadecimal(hexadecimal: String): Boolean = {
    hexadecimal.forall(c => c.isDigit || (c.toUpper >= 'A' && c.toUpper <= 'F'))
  }

  def main(args: Array[String]): Unit = {
    var opcao = 0
    var decimal = 0L
    var hexadecimal = ""
    do {
      println("===============================")
      println(" CONVERSOR DECIMAL-HEXADECIMAL ")
      println("===============================")
      println()
      println("1 - Converter decimal para hexadecimal")
      println("2 - Converter hexadecimal para decimal")
      println("0 - Sair")
      println()
      print("Digite a opcao desejada: ")
      opcao = readLine().toInt
      opcao match {
        case 1 =>
          println()
          print(s"Digite um numero decimal (ate $LimiteMaximoDecimal): ")
          decimal = readLine().toLong
          if (decimal > LimiteMaximoDecimal)
            println("Valor decimal fora do limite permitido.")
          else
            println(s"Valor em hexadecimal: ${DecimalParaHexadecimal(decimal)}")
        case 2 =>
          println()
          print(s"Digite um numero hexadecimal (ate $LimiteMaximoHexadecimal): ")
          hexadecimal = readLine()
          if (!IsValidHexadecimal(hexadecimal))
            println("Valor hexadecimal invalido.")
          else if (HexadecimalParaDecimal(hexadecimal) > LimiteMaximoDecimal)
            println("Valor hexadecimal fora do limite permitido.")
          else
            println(s"Valor em decimal: ${HexadecimalParaDecimal(hexadecimal)}")
        case 0 =>
        case _ => println("Opcao invalida. Tente novamente.")
      }
      println()
    } while (opcao != 0)
  }
}
